### PR TITLE
Adding `DataPollManager` to manage the (periodic) data poll transmission

### PR DESCRIPTION
**Pull request by abtink**
This commit introduces a new class `DataPollManager`. This class
contains all the logic related to sending of the data poll messages
and managing the data poll transmission period.

The data poll interval is derived based on combination of value
from user (from `SetExternalPollPeriod()`) and the internal state
of the OpenThread. Data poll manager may send data polls more
frequently during attach process or when expecting a control-message
from parent or in case of data poll transmission failures or data
poll timeouts.

### CODE REVIEW

File: src/core/thread/data_poll_manager.hpp [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    {
+        kAttachDataPollPeriod  = OPENTHREAD_CONFIG_ATTACH_DATA_POLL_PERIOD,
+        kRetxPollPeiord        = 500,
+        kNoBuffRetxPollPeriod  = 200,
+        kReponseExpectedPeriod = 250,
+    };
+
+    enum
+    {
+        kQuickPollsAfterTimout = 5,   ///< Maximum number of quick data poll tx in case of back-to-back poll timeouts.
```
Comment (LuDuda) [inline src/core/thread/data_poll_manager.hpp]: nit: typo - Timeout?

File: src/core/thread/data_poll_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+    {
+        if ((period == 0) || (period > kAttachDataPollPeriod))
+        {
+            period = kAttachDataPollPeriod;
+        }
+    }
+
+    if (mRetxMode == true)
+    {
+        if ((period == 0) || (period > kRetxPollPeiord))
```
Comment (LuDuda) [inline src/core/thread/data_poll_manager.cpp]: nit: PollPeiord -> PollPeriod

File: src/core/thread/data_poll_manager.hpp (hunk 1 above)
Comment (abtink) [inline src/core/thread/data_poll_manager.hpp]: Thanks... will fix it...

**Comment by abtink** [comment]
Rebased on master to fix merge conflicts with recent merge of 77fed906ac66cb329c7173c15cd0e92395ef3a78 (moving the `otLinkSendDataRequest` to `link_api.cpp`)

**Pull request merged**